pkg/api/vlabs/validate.go

```diff
@@ -1073,30 +1073,84 @@ func validateVMSS(o *OrchestratorProfile, isUpdate bool, storageProfile string)
 	return nil
 }
 
-func (a *AgentPoolProfile) validateWindows(o *OrchestratorProfile, w *WindowsProfile, isUpdate bool) error {
+func (a *Properties) validateWindowsProfile() error {
+	hasWindowsAgentPools := false
+	for _, profile := range a.AgentPoolProfiles {
+		if profile.OSType == Windows {
+			hasWindowsAgentPools = true
+			break
+		}
+	}
+
+	if !hasWindowsAgentPools {
+		return nil
+	}
+
+	o := a.OrchestratorProfile
+	version := ""
+	// This logic is broken because golang cases do not fallthrough by default.
+	// I am leaving this in because I cannot get a clear answer on if we need to continue supporting Swarm + Windows and
+	// RationalizeReleaseAndVersion does not properly handle Swarm.
 	switch o.OrchestratorType {
 	case DCOS:
 	case Swarm:
 	case SwarmMode:
 	case Kubernetes:
-		version := common.RationalizeReleaseAndVersion(
+		version = common.RationalizeReleaseAndVersion(
 			o.OrchestratorType,
 			o.OrchestratorRelease,
 			o.OrchestratorVersion,
-			isUpdate,
+			false,
 			true)
+
 		if version == "" {
 			return errors.Errorf("Orchestrator %s version %s does not support Windows", o.OrchestratorType, o.OrchestratorVersion)
 		}
 	default:
-		return errors.Errorf("Orchestrator %s does not support Windows", o.OrchestratorType)
+		return errors.Errorf("Orchestrator %v does not support Windows", o.OrchestratorType)
 	}
-	if w != nil {
-		if e := w.Validate(o.OrchestratorType); e != nil {
-			return e
+
+	w := a.WindowsProfile
+	if w == nil {
+		return errors.New("WindowsProfile is required when the cluster definition contains Windows agent pools")
+	}
+	if e := validate.Var(w.AdminUsername, "required"); e != nil {
+		return errors.New("WindowsProfile.AdminUsername is required, when agent pool specifies Windows")
+	}
+	if e := validate.Var(w.AdminPassword, "required"); e != nil {
+		return errors.New("WindowsProfile.AdminPassword is required, when agent pool specifies Windows")
+	}
+	if !validatePasswordComplexity(w.AdminUsername, w.AdminPassword) {
```

Comment:
Are these related changes ?

Reply:
Not related - I did some cleanup for how validation is performed for WindowsProfile here.